gh-104584: Baby steps towards generating and executing traces

gvanrossum opened the pull request:
This is mostly so we can benchmark how bad the overhead is.

<!-- gh-issue-number: gh-104584 -->
* Issue: gh-104584
<!-- /gh-issue-number -->

gvanrossum committed:
Boilerplate for new uop optimizer+executor

gvanrossum committed:
Make uop optimizer execute a LOAD_FAST

gvanrossum committed:
Generalize uop optimizer infrastructure

gvanrossum committed:
Give uop_execute an instruction decoding loop

gvanrossum committed:
Make the translator use a switch

gvanrossum committed:
Support LOAD_FAST_LOAD_FAST

gvanrossum committed:
Support LOAD_CONST

gvanrossum committed:
Fix edge case for LOAD_FAST_LOAD_FAST

gvanrossum committed:
Generate SET_IP after each instr

gvanrossum committed:
Turn stats into properties (tp_members)

gvanrossum committed:
Comment out fprintfs

gvanrossum committed:
Enable UOp optimizer by default in main interpreter

gvanrossum committed:
The world isn't ready for 'const int X = 1;'

Use '#define X 1' instead.

gvanrossum committed:
Avoid MSC warning about implicit cast

gvanrossum committed:
Add missing cast in pylifecycle.c

gvanrossum committed:
Generate (some) cases for the executor

gvanrossum committed:
Use write_provenance_header() more

gvanrossum committed:
Recompile flowgraph.c when opcodse_metadata.h changes

gvanrossum committed:
Add uop status to opcode metadata

gvanrossum committed:
Use macros from ceval_macros.h

gvanrossum committed:
Interim -- opcode without oparg using cache entry

gvanrossum committed:
ERROR_IF is no longer a dirty word (but errors crash)

gvanrossum committed:
Regenerate generated_cases.c.h

gvanrossum committed:
Hacky way to avoid flagging macros as uops

(Macros inherit flags from components, but shouldn't inherit IS_UOP.)

gvanrossum committed:
Support DEOPT_IF

gvanrossum committed:
Support ERROR_IF (had to forbid unbound_local_error)

gvanrossum committed:
Attempt to fix check-c-globals

gvanrossum committed:
Add exceptions for static UOp*_Type to c-analyzer

iritkatriel committed:
gh-105481: refactor instr flag related code into a new InstructionFlags class (#105950)

gvanrossum committed:
Add c-analyzer exceptions for real

Revert previous attempts

gvanrossum committed:
More sophisticated uop tracing

gvanrossum committed:
Restore IS_UOP flag generation

gvanrossum committed:
Trace uop optimizations if PYTHON_UOP_DEBUG is set

gvanrossum committed:
Merge remote-tracking branch 'origin/main' into optim-exec

gvanrossum commented on Python/pylifecycle.c:
```diff
@@ -1181,6 +1181,11 @@ init_interp_main(PyThreadState *tstate)
+        PyObject *opt = PyUnstable_Optimizer_NewUOpOptimizer();
+        PyUnstable_SetOptimizer((_PyOptimizerObject *)opt);
+    }
+
```
This is temporary (to root out the bugs, I enable it by default). We should disable it by default and provide a `-X` flag and/or env var to enable it.

gvanrossum commented on Tools/cases_generator/generate_cases.py:
```diff
@@ -290,6 +301,20 @@ def emit_macros(cls, out: Formatter):
+    "INSTRUMENTED_JUMP",
+    "MAKE_FUNCTION_CLOSURE",  # Proxy for 'opcode == SET_FUNCTION_ATTRIBUTE
+)
+
```
This is obviously a work in progress.

- [ ] The error labels are skipped because they call a private function defined in ceval.c. (The problem would go away if we moved the uop optimizer into ceval.c.)
- [x] We can probably just support `tstate`.
- [ ] Ditto `cframe`?
- [ ] I have to look into `kwnames` (what if it's set by regular bytecode but used by a trace, or vice versa?).
- [ ] We need more work for the super-instructions.
- [x] I forget what's the matter with `SET_FUNCTION_ATTRIBUTE`.
- [ ] Most of the others are fundamentally not supported as they manipulate the program counter.

gvanrossum commented on Tools/cases_generator/generate_cases.py:
```diff
@@ -353,7 +379,19 @@ def __init__(self, inst: parser.InstDef):
+                return False
+        return True
+
+    def write(self, out: Formatter, adjust_cache=True) -> None:
```
Should probably rename `adjust_caches` to `executor` or `tier_two` and reverse the sense.

gvanrossum commented on Tools/cases_generator/generate_cases.py:
```diff
@@ -447,9 +491,14 @@ def write_body(self, out: Formatter, dedent: int, cache_adjust: int = 0) -> None
+                    )
+                else:
+                    # Tier 2
+                    assert bits <= 32
```
If this fails, we are trying to fit an object pointer into an oparg. Change the instruction set.

gvanrossum commented on Tools/cases_generator/generate_cases.py:
```diff
@@ -1252,6 +1302,47 @@ def write_instructions(self) -> None:
+                                    if c.name != UNUSED:
+                                        caches.append(c)
+                                    cache_offset += c.size
+                                if len(caches) == 1:
```
This check should be moved into `is_viable_uop()`.

gvanrossum commented on Tools/cases_generator/generate_cases.py:
```diff
@@ -573,16 +622,19 @@ class Analyzer:
     def __init__(self, input_filenames: list[str], output_filename: str,
-                 metadata_filename: str, pymetadata_filename: str):
+                 metadata_filename: str, pymetadata_filename: str,
+                 executor_filename: str):
```
Reformat like black.

gvanrossum reviewed (commented):
Notes to self.

gvanrossum committed:
Move check for oparg-or-one-short-cache into is_viable_uop()

gvanrossum committed:
Support SET_FUNCTION_CLOSURE

gvanrossum committed:
Move uop optimizer to ceval.c so it can use static functions defined there

gvanrossum committed:
Support tstate

We end up forbidding some others: throwflag, exception_unwind,
and (mysteriously, we seem to have to) import_from, import_name.

gvanrossum committed:
Rename 'adjust_cache[s]' to 'tier'

gvanrossum committed:
Reformat Analyzer.__init__ header

gvanrossum committed:
Merge remote-tracking branch 'origin/main' into optim-exec

iritkatriel commented on Tools/cases_generator/generate_cases.py:
```diff
@@ -290,6 +301,28 @@ def emit_macros(cls, out: Formatter):
                 f"(_PyOpcode_opcode_metadata[(OP)].flags & ({name}))")
 
 
+FORBIDDEN_INSTRUCTIONS = (
```
```suggestion
FORBIDDEN_INSTRUCTIONS_IN_UOPS = (
```

gvanrossum replied:
Named it FORBIDDEN_NAMES_IN_UOPS.

markshannon commented on Python/ceval.c:
```diff
@@ -2771,3 +2763,283 @@ void Py_LeaveRecursiveCall(void)
     _Py_LeaveRecursiveCall();
 }
+
+///////////////////// Experimental UOp Interpreter /////////////////////
```
In its own file, please.

gvanrossum replied:
The problem is that there are many instructions that use static functions defined in ceval.c. The rest can go in another file, sure.

markshannon replied:
They can be renamed, as needed.
I'm fairly sure the limited set of instructions we are starting with don't use any static functions.

gvanrossum replied:
> I'm fairly sure the limited set of instructions we are starting with don't use any static functions.

Alas, they do:

- format_awaitable_error
- format_exc_check_arg (4 times)
- format_exc_unbound (3 times)
- format_kwargs_error

markshannon commented on Python/ceval.c:
```diff
@@ -2771,3 +2763,283 @@ void Py_LeaveRecursiveCall(void)
+
+typedef struct {
+    _PyOptimizerObject base;
+    int traces_executed;
```
For stats?
Should be added to `PyStats`

gvanrossum replied:
Okay, I am trying this (see latest commit), but the counter is always printed as 0. What am I missing? I put a `printf()` and the line `OBJECT_STAT_INC(optimization_traces_created)` is definitely executed. Is there another trick I'm missing? (Are there developer docs for pystats?)

markshannon replied:
Stats need to be turned on. They are off by default.
You can save the bother of doing it manually, by using https://github.com/faster-cpython/ideas/actions/workflows/pystats.yml

(Are there developer docs for pystats?)
I don't think so, sorry.

gvanrossum replied:
I figured it out, either pass `-Xpystats` or call `sys._stats_on()` (and `off`). And create `/tmp/py_stats` to get them away from stderr.

markshannon commented on Python/ceval.c:
```diff
@@ -2771,3 +2763,283 @@ void Py_LeaveRecursiveCall(void)
+// UOp opcodes are outside the range of bytecodes or pseudo ops
+#define EXIT_TRACE 512
+#define SET_IP 513
+// TODO: Generate these in Tools/cases_generator
```
Indeed. Put them in `pycore_code.h` for now?

gvanrossum replied:
I'm now generating these, they are put in opcode_metadata.h (which is already generated, so convenient).

markshannon commented on Python/ceval.c:
```diff
@@ -2771,3 +2763,283 @@ void Py_LeaveRecursiveCall(void)
+
+typedef struct {
+    int opcode;
+    int oparg;
```
`intptr_t` for pointer arguments.

markshannon commented on Python/ceval.c:
```diff
@@ -2771,3 +2763,283 @@ void Py_LeaveRecursiveCall(void)
+#define _LOAD_LOCALS 524
+#define _LOAD_FROM_DICT_OR_GLOBALS 525
+
+typedef struct {
```
These should also go in a header file

markshannon commented on Python/ceval.c:
```diff
@@ -2771,3 +2763,283 @@ void Py_LeaveRecursiveCall(void)
+
+typedef struct {
+    _PyExecutorObject executor;  // Base
+    UOpOptimizerObject *optimizer;
```
For stats? Shouldn't be necessary if the stats go in `PyStats`

markshannon commented on Python/ceval.c:
```diff
@@ -2771,3 +2763,283 @@ void Py_LeaveRecursiveCall(void)
+} uop_instruction;
+
+typedef struct {
+    _PyExecutorObject executor;  // Base
```
Maybe name it `base`

markshannon commented on Python/ceval.c:
```diff
@@ -2771,3 +2763,283 @@ void Py_LeaveRecursiveCall(void)
+uop_execute(_PyExecutorObject *executor, _PyInterpreterFrame *frame, PyObject **stack_pointer)
+{
+#ifdef LLTRACE
+    int lltrace = PyDict_Contains(GLOBALS(), &_Py_ID(__lltrace__));
```
This is broken, if we optimize across code objects with globals (stubs and the like).
Maybe just use the environment variable.

markshannon commented on Python/ceval.c:
```diff
@@ -2771,3 +2763,283 @@ void Py_LeaveRecursiveCall(void)
+        oparg = self->trace[pc].oparg;
+#ifdef LLTRACE
+        if (lltrace) {
+            // TODO: Print line numbers; uop names.
```
I'd stick with offsets, rather than line numbers. Line numbers are nice, but are more likely to be misleading if something is wrong.

markshannon commented on Python/ceval.c:
```diff
@@ -2771,3 +2763,283 @@ void Py_LeaveRecursiveCall(void)
+    return frame;
+}
+
+static int
```
This should also go in its own file, and a different file from the uop interpreter.

markshannon commented on Python/ceval.c:
```diff
@@ -2771,3 +2763,283 @@ void Py_LeaveRecursiveCall(void)
+}
+
+static int
+uop_optimize(
```
Yet another file, please.

markshannon commented on Tools/cases_generator/generate_cases.py:
```diff
@@ -242,6 +252,7 @@ class InstructionFlags:
     HAS_CONST_FLAG: bool
     HAS_NAME_FLAG: bool
     HAS_JUMP_FLAG: bool
+    IS_UOP_FLAG: bool = False  # Set later
```
I don't think this is the right way to frame this.
I think it is better to regard all instructions as a list of uops. Many of those lists have length 1, but there is no need to special case it.

gvanrossum replied:
Agreed. I have such a design worked out in my head.

gvanrossum replied:
Done.

markshannon commented on Tools/cases_generator/generate_cases.py:
```diff
@@ -341,6 +374,7 @@ def __init__(self, inst: parser.InstDef):
         self.unmoved_names = frozenset(unmoved_names)
 
         self.instr_flags = InstructionFlags.fromInstruction(inst)
+        self.instr_flags.IS_UOP_FLAG = self.is_viable_uop()
```
See my earlier comment.

markshannon reviewed (commented):
Looks promising.
Don't worry about performance for now, but nice to know it isn't too terrible.

I think we need to break things up into separate files ASAP.

The change from `IS_UOP_FLAG` to instructions being a list of uops can wait for another PR if necessary.

gvanrossum committed:
Auto-generate uop values

gvanrossum committed:
Switch to using macro expansion

(But macros themselves are not yet supported.)

gvanrossum committed:
Remove IS_UOP flag

gvanrossum committed:
Allow pointer-sized ints for Tier-2 oparg

gvanrossum committed:
Rename FORBIDDEN_INSTRUCTIONS

gvanrossum committed:
Remove separate UOpOptimizerObject type

Stop storing stats in the optimize.

gvanrossum committed:
Try to use pystats

gvanrossum committed:
Rationalize errors

PYTHON_UOP_DEBUG=
0: no debugging
1: print creation of non-empty traces only
2: all details about trace creation, + entering traces an deopt/error exits
3: all details about trace creation and execution

gvanrossum committed:
Fix typo in pystats output

PS. I finally understand wht I was getting zeros:
You need to either pass -Xpystats, or call sys._stats_on().
The confusing part is that the output gets printed even if it's off.
And to get the output not on stderr, create /tmp/py_stats.

gvanrossum committed:
Move uop_optimizer back to optimizer.c

Keep the executor in ceval.c; it depends on static functions there.
For the shared definitions I created a new file, pycore_uops.h.
Renamed the definitions to have a _Py prefix.

gvanrossum committed:
Fix 4/6 failing tests

gvanrossum committed:
Receiving end of macro expansion

gvanrossum committed:
Avoid warnings; subtler debug output levels

gvanrossum committed:
Add eval breaker check to _PyUopExecute; fixes test_threading

gvanrossum committed:
Merge remote-tracking branch 'origin/main' into optim-exec

gvanrossum committed:
Add opcode name to ADD_TO_TRACE debug output

gvanrossum committed:
Fix test_memoryview again by excluding BEFORE_WITH

Apparently the fix using SET_LOCALS_FROM_FRAME() isn't enough
(it fixes some but not all failures). Possibly the problem is due
to having *two* BEFORE_WITH uops in one trace?

Unclear what's wrong with that opcode.

gvanrossum committed:
Fix test_memoryview by excluding BEFORE_WITH

Unclear what's wrong with that opcode.

gvanrossum committed:
Fix Windows test_mmap -- by removing BUILD_SLICE/STORE_SLICE

gvanrossum committed:
Fix bug in ENTER_EXECUTOR - on error go to resume_with_error

Now test_mmap on Windows passes,
and hopefully so does test_threading on UNIX.

gvanrossum committed:
Silence warnings on Windows

Instead of `int oparg` we now have `uint64_t operand`.
In the Tier 2 interpreter, we add `int oparg` and set it
to a truncated version of `operand` so instructions can
reference it.

gvanrossum committed:
Replace oparg with operand in more debug output

gvanrossum committed:
Add -Xuops flag and debug tweaks

The uops optimizer is enabled in the main interpreter when either:
- -Xuops is passed; or
- PYTHONUOPS=1 is set.

The env var for uops debugging is now called PYTHONUOPSDEBUG.
It only has an effect when configured --with-pydebug *and* the uops
optimizer is enabled (see above).

Also fixed a compiler warning about %ld with an int64 arg.

gvanrossum committed:
Revert "Fix 4/6 failing tests"

Those tests now only fail when run with -Xuops.
We should probably fix the uop executor instead.
Also, test_embed now passes even with -Xuops.

This reverts commit 91dd86cc64fc450e5cec876514e0439bd81cfcd1.

gvanrossum committed:
Use PRIu64 instead of %lld/%llu formats

kumaraditya303 commented on Python/optimizer.c:
```diff
@@ -278,3 +280,201 @@ PyUnstable_Optimizer_NewCounter(void)
+        return -1;
+    }
+    executor->base.execute = _PyUopExecute;
+    Py_INCREF(self);
```
This incref looks unnecessary

gvanrossum replied:
Very good. I owe you!

gvanrossum committed:
Remove superfluous INCREF -- hope it fixes the refleaks

blurb-it[bot] committed:
📜🤖 Added by blurb_it.
